### PR TITLE
Respect dist autograd context in torch.jit._fork.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34360 Respect dist autograd context in torch.jit._fork.**

The distributed autograd context sets up a thread local context id
which is used to perform appropriate book keeping and autograd recording of RPC
functions in the forward pass.

However, if we use torch.jit._fork within the distributed autograd context, the
code executed within torch.jit._fork will lose this context since it is run in
a separate JIT thread and the thread local is not set in that thread.

To fix this problem, we pass in the distributed autograd context to
torch.jit._fork similar to what we did in
https://github.com/pytorch/pytorch/pull/16101.

Closes https://github.com/pytorch/pytorch/issues/33479

Differential Revision: [D20301352](https://our.internmc.facebook.com/intern/diff/D20301352/)

### CODE REVIEW

**Review by wanchaol — COMMENTED** [review]
The approach looks good to me. But I am a bit concerned why we are keeping the `current_context_id` as thread_local? Looking at DistributedContainer I think we could make it as a static member of the DistributedContainer singleton, and this issue should just go away right?

**Comment by pritamdamania87** [comment]
> The approach looks good to me. But I am a bit concerned why we are keeping the current_context_id as thread_local? Looking at DistributedContainer I think we could make it as a static member of the DistributedContainer singleton, and this issue should just go away right?

Having `current_context_id` as a thread local was by design. If we have multiple JIT threads running independent forward and backward passes, each one of them needs their own unique `current_context_id` to ensure we have independent autograd graphs for each training loop for distributed autograd.

File: torch/csrc/jit/runtime/interpreter.cpp
Comment (zdevito) [inline torch/csrc/jit/runtime/interpreter.cpp]: ifdef for distributed should not appear in interpreter.cpp, this is a sign of strong coupling. I guess an abstraction where we introduce a `SavedThreadLocalState` object that holds all the pytorch-specific thread local state that interpreter needs to save and restore when it suspends. Then we have public functions that the interpreter uses in interpreter continuations:
```
unique_ptr<SavedThreadLocalState> saveThreadLocalState();
restoreThreadLocalState(const SavedThreadLocalState&);
```

**Comment by ilia-cher** [comment]
we have `DebugInfoGuard` guard and `ThreadLocalDebugInfoBase' storage structs, their goal is to pass some information along model execution (both fwd and backward) and we also ensure that  whenever you call 'fork' or go into autograd threads we pass this info across threads too and clean up after.
So, it seems to be a good candidate in general to store this kind of data?

File: torch/csrc/jit/runtime/interpreter.cpp
Comment (ilia-cher) [inline torch/csrc/jit/runtime/interpreter.cpp]: instead of modifying the interpreter, you could reuse, e.g. https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/ParallelThreadPoolNative.cpp#L73 and
https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/engine.cpp#L515 - some update (e.g. name) is needed for ThreadLocalDebugInfoBase, since it's not going to be only for debug info;
Btw, we've just coincidentally had a discussion re. this update, and, e.g. thought about the ability to pass some arbitrary data from higher-level python layers (e.g. logical name of distributed node) too

**Comment by pritamdamania87** [comment]
Implemented @zdevito's proposal in this PR to save and restore the thread local state.

@ilia-cher I looked into the ThreadLocalDebugInfoBase structs as you suggested. Although, I felt it was probably not the best route for this change for the following reasons:

1. The API is intended for end users to use (using setThreadLocalDebugInfo) and we don't want users to mess with the thread locals for grad_mode and dist_autograd_context_id.
2. I'm not sure if we want to propagate these thread locals to all threads via at::launch. Feels safe to have them enabled only in JIT continuations for now.
3. Mixing user specified debug info and internal thread local state in a single structure didn't seem right.

**Comment by pritamdamania87** [comment]
cc @satgera @jjlilley

File: torch/csrc/ThreadLocalState.h
Comment (jjlilley) [inline torch/csrc/ThreadLocalState.h]: Wondering aloud: Do we really need to wrap in a shared ptr, if this is [currently] just an {int64,int32}?

I understand why this makes sense (especially if we add fields in future), but it also feels expensive to add an extra malloc + free, plus several atomic ops, for each jit.fork/wait.

**Comment by ilia-cher** [comment]
> Implemented @zdevito's proposal in this PR to save and restore the thread local state.
> 
> @ilia-cher I looked into the ThreadLocalDebugInfoBase structs as you suggested. Although, I felt it was probably not the best route for this change for the following reasons:
> 
> 1. The API is intended for end users to use (using setThreadLocalDebugInfo) and we don't want users to mess with the thread locals for grad_mode and dist_autograd_context_id.
> 2. I'm not sure if we want to propagate these thread locals to all threads via at::launch. Feels safe to have them enabled only in JIT continuations for now.
> 3. Mixing user specified debug info and internal thread local state in a single structure didn't seem right.

A lot of thread local variables need to be propagated across fork and at::parallel_for, what are the examples of variables that do not need to be propagated?

**Comment by ilia-cher** [comment]
My point is - we do want to propagate across fork and at::parallel_for and autograd threads, by default, so why not consolidate in one thread local data struct that would guarantee that properties are propagated across threads?

**Comment by pritamdamania87** [comment]
> A lot of thread local variables need to be propagated across fork and at::parallel_for, what are the examples of variables that do not need to be propagated?

The `dist_autograd_context_id` is only used in the forward pass and doesn't need to be propagated to autograd threads.

> My point is - we do want to propagate across fork and at::parallel_for and autograd threads, by default, so why not consolidate in one thread local data struct that would guarantee that properties are propagated across threads?

I'm a little concerned about implicitly propagating thread_locals across any arbitrary at::launch(), given that we don't know what that thread is doing and how the thread_locals might affect its behavior.

File: torch/csrc/ThreadLocalState.h
Comment (pritamdamania87) [reply torch/csrc/ThreadLocalState.h]: Good point, I'll avoid the shared_ptr for now and we can revisit this later when we add more fields.

**Review by jjlilley — COMMENTED** [review]
fwiw, this looks good to me now that the alloc + atomic ops are gone.
@zdevito might have more comments?

File: torch/csrc/jit/runtime/interpreter.cpp
Comment (gqchen) [inline torch/csrc/jit/runtime/interpreter.cpp]: we can remove this line?

File: torch/csrc/jit/runtime/interpreter.cpp
Comment (gqchen) [inline torch/csrc/jit/runtime/interpreter.cpp]: this line as well?

**Review by gqchen — COMMENTED** [review]
two minor comments. Otherwise, it looks good to me.

**Review by zdevito — APPROVED** [review]
Looks good. Thanks for refactoring this!

**Comment by jjlilley** [comment]
Pritam, this looks good to me as well, but can you hold off landing until the pending rollback lands?

We lost 500+ tests (rpc and autograd related) yesterday afternoon, the rollback seems to be landing, but would be good to test without the tests all disabled.

File: torch/testing/_internal/distributed/rpc/jit/dist_autograd_test.py
Comment (xush6528) [inline torch/testing/_internal/distributed/rpc/jit/dist_autograd_test.py]: To silent linter, need to annotate `def remote_add(t1, t2, dst: str):  # noqa: E999`

**Comment by pritamdamania87** [comment]
As per Jeremy's suggestion, will hold off on landing this until things stabilize in trunk.

**Pull request closed without merging**
